crates/red_knot_workspace/src/workspace/metadata.rs

```diff
@@ -1,67 +1,191 @@
+                                        second: package.root().to_path_buf(),
+                                    });
+                                }
+
+                                if package.root() == workspace_root {
+                                    workspace_package = Some(package);
+                                } else if !package.root().starts_with(workspace_root) {
+                                    return Err(WorkspaceDiscoveryError::PackageOutsideWorkspace {
+                                        package_name: package.name().clone(),
+                                        package_root: package.root().to_path_buf(),
+                                        workspace_root: workspace_root.to_path_buf(),
+                                    });
+                                }
+                            }
+
+                            let workspace_package = workspace_package
+                                .expect("workspace package to be part of the workspace's packages");
+
+                            let settings = workspace_package
+                                .configuration
+                                .to_workspace_settings(workspace_root, &packages);
+
+                            packages.sort_unstable_by(|a, b| a.root().cmp(b.root()));
+
+                            return Ok(Self {
+                                root: workspace_root.to_path_buf(),
+                                packages,
+                                settings,
+                            });
+                        }
+                        CollectedPackagesOrStandalone::Standalone(package) => {
+                            closest_package = Some(package);
+                            break;
+                        }
+                    }
+                }
+
+                // Not a workspace itself, keep looking for an enclosing workspace.
+                if closest_package.is_none() {
+                    closest_package = Some(package);
+                }
+            }
         }
 
-        // TODO: Respect the package configurations when resolving settings (e.g. for the target version).
-        let settings = configuration.into_workspace_settings(&root);
+        // No workspace found, but maybe a pyproject.toml was found.
+        let package = if let Some(enclosing_package) = closest_package {
+            tracing::debug!("Single package workspace at '{}'", enclosing_package.root());
 
-        let workspace = WorkspaceMetadata {
-            root,
-            packages: vec![package],
-            settings,
+            enclosing_package
+        } else {
+            tracing::debug!("The ancestor directories contain no `pyproject.toml`. Falling back to a virtual project.");
+
+            // Create a package with a default configuration
+            PackageMetadata {
+                name: path.file_name().unwrap_or("<virtual>").into(),
+                root: path.to_path_buf(),
```

Comment:
I think we could consider adding a feature like this if it comes up often. Supporting it would complicate things a bit, because we then create a workspace that covers more files than `path`, that means that red knot needs to filter out all files that are outside `path` before calling `check`.